### PR TITLE
Scheduler enhancements

### PR DESCRIPTION
**Pull request by mfix22**
Implements new scheduler capabilities:
- Create, edit, and delete tags
- Apply tags to scheduled queries
- Display last execution data: row count, duration, completed at, and status/error
- Display when scheduled query will run next
- Filter scheduled queries by all of the above in the search bar

Closes https://github.com/plotly/falcon/issues/492

### CODE REVIEW

**Comment by nicolaskruchten** [comment]
Starting with a UI-level pass, then will drill into the code :)

Overall this is quite nice, great job guys!

A few oddities...

- [ ] I initially had some problems with "run now" where even for trivial queries it was extremely slow, but this appeared to be on the 'saving' step and hard to replicate.

- [ ] If I leave the modal open after a manual run, the "scheduled to run" gets negative:

image

- [ ] after hitting "run now" the "last ran" in the modal isn't up to date

- [ ] When hitting "run now" on a query saved by another user, I got this...

image

- [ ] "run now" while not logged in doesn't seem to work... just hangs on "saving..."

- [ ] why is the query list box so small in the schedule tab ? I can only see 4 queries even though I'm on a vertical monitor and have lots of whitespace in the window.

- [ ] could we have "last run" mean "last run started" instead of "last run ended" such that sorting by most-recently-run puts "currently running" at the top? Or maybe provide both options?

- [ ] possible to have 'next run time' in the list view, and be able to sort by next run time?

**Comment by nicolaskruchten** [comment]
[ ] In the modal, could we see absolute times as well as relative times? Ideally we would see absolute start and end times for the "last ran" (possibly as a tooltip on the relative time?) and the next scheduled absolute time. This is pretty important operationally and especially when correlating errors to outages or database problems!

**Comment by nicolaskruchten** [comment]
- [ ] right now we have filters for "success" and "error". "running" seems to be rolled into "success" but could we break it out into its own button?

**Comment by nicolaskruchten** [comment]
Looks like you guys checked off all the boxes, thanks very much for the quick turnaround! Is this ready for another QA/code-level pass from me?

**Comment by mfix22** [comment]
@nicolaskruchten thanks for catching all those issues! Should be ready for another pass 👍

File: app/components/Settings/OptionsDropdown/sql-options.jsx [hunk 1]
```diff
@@ -33,7 +33,7 @@ export default class SQLOptions extends Component {
             return (
                 <div>
                     <div>{'Hm.. there was an error loading up your tables'}</div>
-                    <div style={{color: 'red'}}>{JSON.stringify(tablesRequest)}</div>
+                    <div style={{color: '#EF553B'}}>{JSON.stringify(tablesRequest)}</div>
```
Comment (shannonlal) [inline app/components/Settings/OptionsDropdown/sql-options.jsx]: Just a thought.  Maybe define '#EF553B' as a constant. Not a requirements

File: app/components/Settings/scheduler/modals/create-modal/create-modal.jsx [hunk 2]
```diff
@@ -109,29 +143,30 @@ class CreateModal extends Component {
...
+                name: this.state.name ? this.state.name.trim() : '',
+                tags: this.state.tags.map(t => t.id)
             })
-            .then((res) => {
+            .then(res => {
                 this.setState({
-                  successMessage: 'Scheduled query saved successfully!',
-                  saving: false,
-                  datasetUrl: getDatasetURL(res.fid)
+                    successMessage: 'Scheduled query saved successfully!',
```
Comment (shannonlal) [inline app/components/Settings/scheduler/modals/create-modal/create-modal.jsx]: Just maybe fix spacing issues with previous lines

Comment (nicolaskruchten) [reply app/components/Settings/scheduler/modals/create-modal/create-modal.jsx]: let's ignore all spacing/formatting issues for now, as we'll run `prettier` on the entire codebase soon :)

File: app/components/Settings/OptionsDropdown/sql-options.jsx (hunk 1 above)
Comment (briandennis) [reply app/components/Settings/OptionsDropdown/sql-options.jsx]: Good catch! Updated ✔️

**Comment by briandennis** [comment]
@nicolaskruchten @shannonlal  This should be up to date with your feedback. Is there anything else you want to review before we merge?

**Comment by nicolaskruchten** [comment]
yes I'm still reviewing today :)

File: test/backend/routes.tags.spec.js [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+                .then(assertResponseStatus(200))
+                .then(getResponseJson)
+                .then(json => {
+                    assert.equal(json.length, 2);
+                    assert.equal(json[0].name, TAG.name);
+                    assert.equal(json[1].name, TAG_2.name);
+                });
+        });
+
+        // TODO: run this test once tags are persisted on queries
```
Comment (nicolaskruchten) [inline test/backend/routes.tags.spec.js]: Could you please do a quick pass and make sure all TODOs are either deleted or moved to a separate issue (one issue with a checklist would be fine) ?

**Comment by nicolaskruchten** [comment]
As a process note, I think I would have found it easier/quicker to review this if we had split up this PR into two parts: tags management in one PR and last-run info in another :)

File: test/backend/routes.tags.spec.js (hunk 3 above)
Comment (briandennis) [reply test/backend/routes.tags.spec.js]: Are you referring to TODOs from our work? I believe that was the only one and I removed it as it's outdated.

**Comment by briandennis** [comment]
1. If you delete a query while it's running, it will be rewritten (undoing the delete) upon completion. As you mentioned, not new behavior, but the shortcoming may be more noticeable now. 

2. Old-Falcon should correctly interpret the New-Falcon files. Similar to the last iteration where we added query names, new query metadata will simply be ignored by old clients. I'm trying to manually test this for you, but am having some trouble building master. I'll let you know when I've confirmed  👍

**Comment by briandennis** [comment]
re process note: Good to know! Going forward, we can break these up more and run our plans by you before we begin.

File: test/backend/routes.tags.spec.js (hunk 3 above)
Comment (nicolaskruchten) [reply test/backend/routes.tags.spec.js]: cool, thanks! it's the only one I saw but I just wanted a check :)

**Comment by nicolaskruchten** [comment]
re 1 & 2: 👍 

💃 once case 2 has been checked :)

**Comment by briandennis** [comment]
Confirmed Falcon `master` correctly reads queries configured from this branch. Some metadata is lost (namely tags), but that's expected as `master` knows nothing about them.

**Comment by nicolaskruchten** [comment]
Awesome, thanks team!

**Pull request merged**